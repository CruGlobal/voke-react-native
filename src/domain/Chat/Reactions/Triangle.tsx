import React, { ReactElement } from 'react';
import { View, ViewStyle } from 'react-native';

interface Props {
  color?: string;
  rotate?: number;
  size?: number;
  style?: ViewStyle;
}

const Triangle = (props: Props): ReactElement => {
  const { color = '#fff', rotate = 0, size = 1, style = {} } = props;

  const chevronTriangle = {
    backgroundColor: 'transparent',
    borderWidth: 0,
    borderTopWidth: 5 * size,
    borderLeftWidth: 5 * size,
    borderColor: 'transparent',
    borderLeftColor: color,
  };

  return (
    <View
      style={[
        {
          width: 10 * size,
          height: 10 * size,
          transform: [{ rotate: `${rotate}deg` }],
        },
        style,
      ]}
    >
      <View
        style={[
          chevronTriangle,
          {
            position: 'absolute',
            bottom: 0,
            left: 0,
            transform: [{ scale: -1 }],
          },
        ]}
      />
      <View
        style={[
          chevronTriangle,
          {
            position: 'absolute',
            bottom: 0,
            right: 0,
            transform: [{ scaleY: -1 }],
          },
        ]}
      />
    </View>
  );
};

export default Triangle;
