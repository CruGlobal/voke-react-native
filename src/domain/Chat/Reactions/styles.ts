import { StyleSheet } from 'react-native';
import theme from 'utils/theme';

const styles = StyleSheet.create({
  popupWrapper: {
    position: 'absolute',
    width: '100%',
    top: -76,
  },
  popup: {
    backgroundColor: '#fefefe',
    height: 60,
    borderRadius: theme.radius.l,
    paddingHorizontal: theme.spacing.s,
    justifyContent: 'space-evenly',
    flexDirection: 'row',
  },
  popupTail: {
    position: 'absolute',
    bottom: -10,
    left: 14,
    zIndex: -1,
  },
  emoji: {
    flex: 1,
    height: 60,
    textAlign: 'center',
    justifyContent: 'center',
    alignItems: 'stretch',
    alignSelf: 'stretch',
  },
  emojiChar: {
    lineHeight: 60,
    fontSize: 38,
    textAlign: 'center',
    color: '#000',
  },
});

export default styles;
