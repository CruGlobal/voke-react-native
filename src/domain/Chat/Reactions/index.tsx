import React from 'react';
import { Pressable, Text, View } from 'react-native';
import * as Animatable from 'react-native-animatable';

import Triangle from './Triangle';
import styles from './styles';

interface Props {
  onReaction: (newReaction: string) => void;
}

const Reactions = ({ onReaction }: Props): React.ReactElement => {
  const emojis = ['♥️', '😄', '😂', '🤗', '😓'];
  const onReactionSelected = (reaction: string): void => {
    onReaction(reaction);
  };
  return (
    <Animatable.View
      animation="fadeInUp"
      duration={500}
      delay={120}
      easing={'ease-in-out-cubic'}
      useNativeDriver={process.env.JEST_WORKER_ID ? false : true}
      style={styles.popupWrapper}
      testID="reactionsPopup"
    >
      <View style={styles.popup}>
        {emojis.map((emoji, index) => (
          <Pressable
            key={index}
            style={styles.emoji}
            onPressIn={(): void => {
              onReactionSelected(emoji);
            }}
            testID={'reaction-' + index}
          >
            <Text style={styles.emojiChar}>{emoji}</Text>
          </Pressable>
        ))}
      </View>
      <Triangle style={styles.popupTail} size={4} color="#eee" />
    </Animatable.View>
  );
};

export default Reactions;
