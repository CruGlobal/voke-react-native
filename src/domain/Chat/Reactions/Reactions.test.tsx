import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';

import Reactions from '../index';

it('renders correctly', () => {
  const { queryByText, toJSON } = render(
    <Reactions
      onReaction={(): void => {
        // void.
      }}
    />,
  );

  expect(toJSON()).toMatchSnapshot();
  expect(queryByText(/♥️/i)).toBeTruthy();
  expect(queryByText(/😄/i)).toBeTruthy();
  expect(queryByText(/😂/i)).toBeTruthy();
  expect(queryByText(/🤗/i)).toBeTruthy();
  expect(queryByText(/😓/i)).toBeTruthy();
});

it('emit events correctly', () => {
  const onReaction = jest.fn();
  const { getByTestId } = render(
    <Reactions
      onReaction={(reaction): void => {
        onReaction(reaction);
      }}
    />,
  );

  fireEvent(getByTestId('reaction-0'), 'onPressIn');
  expect(onReaction).toHaveBeenCalledWith('♥️');

  fireEvent(getByTestId('reaction-1'), 'onPressIn');
  expect(onReaction).toHaveBeenCalledWith('😄');

  fireEvent(getByTestId('reaction-2'), 'onPressIn');
  expect(onReaction).toHaveBeenCalledWith('😂');

  fireEvent(getByTestId('reaction-3'), 'onPressIn');
  expect(onReaction).toHaveBeenCalledWith('🤗');

  fireEvent(getByTestId('reaction-4'), 'onPressIn');
  expect(onReaction).toHaveBeenCalledWith('😓');
});
