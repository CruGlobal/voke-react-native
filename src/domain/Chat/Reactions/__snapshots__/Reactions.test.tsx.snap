// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  forwardedRef={[Function]}
  style={
    Object {
      "opacity": 0,
      "position": "absolute",
      "top": -76,
      "transform": Array [
        Object {
          "translateY": 100,
        },
      ],
      "width": "100%",
    }
  }
  testID="reactionsPopup"
>
  <View
    style={
      Object {
        "backgroundColor": "#fefefe",
        "borderRadius": 12,
        "flexDirection": "row",
        "height": 60,
        "justifyContent": "space-evenly",
        "paddingHorizontal": 8,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "stretch",
          "alignSelf": "stretch",
          "flex": 1,
          "height": 60,
          "justifyContent": "center",
          "textAlign": "center",
        }
      }
      testID="reaction-0"
    >
      <Text
        style={
          Object {
            "fontSize": 38,
            "lineHeight": 60,
            "textAlign": "center",
          }
        }
      >
        ♥️
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "stretch",
          "alignSelf": "stretch",
          "flex": 1,
          "height": 60,
          "justifyContent": "center",
          "textAlign": "center",
        }
      }
      testID="reaction-1"
    >
      <Text
        style={
          Object {
            "fontSize": 38,
            "lineHeight": 60,
            "textAlign": "center",
          }
        }
      >
        😄
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "stretch",
          "alignSelf": "stretch",
          "flex": 1,
          "height": 60,
          "justifyContent": "center",
          "textAlign": "center",
        }
      }
      testID="reaction-2"
    >
      <Text
        style={
          Object {
            "fontSize": 38,
            "lineHeight": 60,
            "textAlign": "center",
          }
        }
      >
        😂
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "stretch",
          "alignSelf": "stretch",
          "flex": 1,
          "height": 60,
          "justifyContent": "center",
          "textAlign": "center",
        }
      }
      testID="reaction-3"
    >
      <Text
        style={
          Object {
            "fontSize": 38,
            "lineHeight": 60,
            "textAlign": "center",
          }
        }
      >
        🤗
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "stretch",
          "alignSelf": "stretch",
          "flex": 1,
          "height": 60,
          "justifyContent": "center",
          "textAlign": "center",
        }
      }
      testID="reaction-4"
    >
      <Text
        style={
          Object {
            "fontSize": 38,
            "lineHeight": 60,
            "textAlign": "center",
          }
        }
      >
        😓
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "height": 40,
          "transform": Array [
            Object {
              "rotate": "0deg",
            },
          ],
          "width": 40,
        },
        Object {
          "bottom": -10,
          "left": 14,
          "position": "absolute",
          "zIndex": -1,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderColor": "transparent",
            "borderLeftColor": "#eee",
            "borderLeftWidth": 20,
            "borderTopWidth": 20,
            "borderWidth": 0,
          },
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "transform": Array [
              Object {
                "scale": -1,
              },
            ],
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderColor": "transparent",
            "borderLeftColor": "#eee",
            "borderLeftWidth": 20,
            "borderTopWidth": 20,
            "borderWidth": 0,
          },
          Object {
            "bottom": 0,
            "position": "absolute",
            "right": 0,
            "transform": Array [
              Object {
                "scaleY": -1,
              },
            ],
          },
        ]
      }
    />
  </View>
</View>
`;
